azkaban-web-server/src/main/java/cloudflow/services/FlowServiceImpl.java

```diff
@@ -0,0 +1,200 @@
+    }
+
+    return flow.get();
+  }
+
+
+  private String extractArgumentValue(String argName, String[] argValues) {
+    /* Currently we support only one value for each of the parameters and return an error
+    otherwise.
+    */
+    requireNonNull(argName);
+    requireNonNull(argValues);
+    if (argValues.length != 1) {
+      throw new RuntimeException(format("Argument %s should have exactly one value", argName));
+    }
+    return argValues[0];
+  }
+
+
+  private void setFlowCreateAndLastModifiedInfo(FlowResponse flow) {
+    /* Get first flow version to find when the flow was created */
+    Optional<FlowVersion> firstFlowVersion = flowVersionDao.getVersion(flow.getId(), 1);
+    if (!firstFlowVersion.isPresent()) {
+      throw new CloudFlowNotFoundException("First version of flow: " + flow.getId() + "does not "
+          + "exist");
+    }
+    flow.setCreatedByUser(firstFlowVersion.get().getCreatedBy());
+    flow.setCreatedOn(firstFlowVersion.get().getCreateTime());
+
+    /* Get the last flow version that was uploaded as part of the project to extract modifiedBy
+     and modifiedOn */
+    Optional<FlowVersion> lastFlowVersion = flowVersionDao.getVersion(flow.getId(),
+        flow.getFlowVersionCount());
+    if (!lastFlowVersion.isPresent()) {
+      throw new CloudFlowNotFoundException("Latest flow version of flow: " + flow.getId() + "does"
+          + " not exist");
+    }
+    flow.setModifiedByUser(lastFlowVersion.get().getCreatedBy());
+    flow.setModifiedOn(lastFlowVersion.get().getCreateTime());
+  }
+
+
+  private void setLastFlowVersion(FlowResponse flow) {
+    Optional<FlowVersion> lastFlowVersion = flowVersionDao.getVersion(flow.getId(),
+        flow.getFlowVersionCount());
+    if (!lastFlowVersion.isPresent()) {
+      throw new CloudFlowNotFoundException("Latest flow version of flow: " + flow.getId() + "does"
+          + "not exist");
+    }
+    flow.setLastVersion(lastFlowVersion.get());
+  }
+
+
+  @Override
+  public List<FlowResponse> getAllFlows(User user, Map<String, String[]> queryParamMap) {
+    Optional<String> projectId = Optional.empty();
+    Optional<String> projectVersion = Optional.empty();
+    Optional<String> projectName = Optional.empty();
+
+    for (Map.Entry<String, String[]> entry : queryParamMap.entrySet()) {
+      String key = entry.getKey();
+      String[] value = entry.getValue();
+      if (key.equals(PROJECT_ID_PARAM)) {
+        projectId = Optional.of(extractArgumentValue(key, value));
+      } else if (key.equals(PROJECT_VERSION_PARAM)) {
+        projectVersion = Optional.of(extractArgumentValue(key, value));
+      } else if (key.equals(PROJECT_NAME_PARAM)) {
+        projectName = Optional.of(extractArgumentValue(key, value));
+      }
+    }
+
+    /* Both projectId and projectName cannot be passed as query params */
+    if (projectId.isPresent() && projectName.isPresent()) {
+      throw new CloudFlowException("Filtering by both project id and project name is not "
```

Comment:
Perhaps `CloundFlowValidationException` instead ?

Reply:
I was debating between the two, changed to CloundFlowValidationException.